src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -170,22 +180,25 @@ public virtual void Abort()
 
                                     _cachedPingMessage = Protocol.WriteToArray(PingMessage.Instance);
 
-                                    Log.UsingHubProtocol(_logger, Protocol.Name);
-
                                     UserIdentifier = userIdProvider.GetUserId(this);
 
                                     if (Features.Get<IConnectionInherentKeepAliveFeature>() == null)
                                     {
                                         // Only register KeepAlive after protocol negotiated otherwise KeepAliveTick could try to write without having a ProtocolReaderWriter
-                                        Features.Get<IConnectionHeartbeatFeature>()?.OnHeartbeat(state => ((HubConnectionContext)state).KeepAliveTick(), this);
+                                        Features.Get<IConnectionHeartbeatFeature>()?.OnHeartbeat(state => ((HubConnectionContext) state).KeepAliveTick(), this);
                                     }
 
+                                    Log.NegotiateComplete(_logger, Protocol.Name);
+                                    await WriteNegotiateResponseAsync(new NegotiationResponseMessage(null));
                                     return true;
                                 }
                             }
                             else if (result.IsCompleted)
                             {
-                                break;
+                                // connection was closed before we every received a response
+                                // can't send a negotiate response because there is no longer a connection
+                                Log.NegotiateFailed(_logger, null);
```

Comment:
Don't we log this on the outside?

Reply:
There is nothing in `OnConnectedAsync` which is the only caller of this method. Where else would it be?